python/fusion_engine_client/analysis/data_loader.py

```diff
@@ -507,6 +510,103 @@ def get_index(self):
     def get_input_path(self):
         return self.reader.input_file.name
 
+    def convert_to_p1_time(self,
+                           times: Union[Iterable[Union[datetime, gpstime, Timestamp, float]],
+                                        Union[datetime, gpstime, Timestamp, float]],
+                           assume_utc: bool = False) ->\
+            np.ndarray:
+        """!
+        @brief Convert UTC or GPS timestamps to P1 time.
+
+        @param times A list of one or more timestamps to be converted, using any of the following formats:
+               - `datetime` - A UTC or local timezone date and time
+               - `gpstime` - A GPS timestamp
+               - A @ref fusion_engine_client.messages.timestamps.Timestamp containing GPS time or P1 time
+               - A @ref fusion_engine_client.messages.timestamps.MeasurementTimestamps containing GPS time or P1 time
+               - `float` - A GPS or P1 time value (in seconds)
+                 - Note that UTC timestamps cannot be specified `float` unless `assume_utc == True`
+        @param assume_utc If `True`, assume all `float` values greater than the POSIX offset for 2000/1/1 are UTC
+               timestamps in seconds
+
+        @return A numpy array containing P1 time values (in seconds), or `nan` if the value could not be converted.
+        """
+        # Load pose messages, which contain the relationship between P1 and GPS time. Omit any NAN values from the
+        # reference timestamps.
+        #
+        # If no pose data is available, we can't convert to P1 time. We'll return nan for any values that are not
+        # already P1 time below.
+        if PoseMessage.MESSAGE_TYPE in self.data:
+            # If the caller already read in pose data, we'll use the cached messages read with whatever parameters they
+            # specified (e.g., a specific time range). That way, A) we don't read from disk multiple times, and B) for a
+            # very long log, we don't read a ton of data from disk when the caller is only interested in a short
+            # snippet. We assume the requested timestamps won't be too far out of the time range specified for the
+            # cached messages.
+            self.logger.debug('Using existing cached pose data for P1 time conversion.')
+            pose_data = self.data[PoseMessage.MESSAGE_TYPE]
+        else:
+            result = self.read(message_types=[PoseMessage], return_numpy=True)
+            pose_data = result[PoseMessage.MESSAGE_TYPE]
+
+        idx = ~np.logical_or(np.isnan(pose_data.p1_time), np.isnan(pose_data.gps_time))
```

Comment:
In what scenario would P1 time be NAN while GPS time is non-nan, or vice versa? I'm guessing it has something to do with initialization.

Reply:
GPS time can be nan if you start up in a place without GPS, like inside a parking garage. Ideally P1 time should never be nan, but it's always good to check just to be safe.